Introduce Storage Connections management dialog

mkemel opened the pull request:
Storage Connections management dialog is opened by clicking the
Connections button in the Storage Domains table when iSCSI Storage
Domain is selected.

The dialog displays storage connections, attached to the domain, and
also has a switch to show all the iSCSI connections. It allows 5
actions to be performed on the connections:
1. Add new connection
2. Edit connection
3. Remove connection
4. Attach connection to the domain
5. Detach connection from the domain
Which provides UI for the existing storage connections API
Live changes are not supported, thus the domain should be in
maintenance mode to perform actions (4),(5), and also (2),(3) for
attached connections

Bug: https://bugzilla.redhat.com/show_bug.cgi?id=977379
![Screenshot from 2022-02-15 14-24-01](https://user-images.githubusercontent.com/87977971/154064140-c1f8ee02-88b1-4931-a99b-d6cd84fc5b5b.png)
![Screenshot from 2022-02-15 14-25-37](https://user-images.githubusercontent.com/87977971/154064143-8ed3d261-8468-4ff3-8e45-52159e601d1a.png)
![Screenshot from 2022-02-15 14-26-03](https://user-images.githubusercontent.com/87977971/154064145-35dd3f9c-7aac-4ab7-9506-efebc157db7c.png)
![Screenshot from 2022-02-15 14-26-22](https://user-images.githubusercontent.com/87977971/154064149-71be4610-841a-47aa-bcef-7dd304a33a45.png)
![Screenshot from 2022-02-15 14-26-40](https://user-images.githubusercontent.com/87977971/154064150-298fc8e9-9f10-4c58-98d0-76a3c0061662.png)
.

bennyz commented on src/integrations/buttons.js:
```diff
@@ -140,6 +141,29 @@ function addHostCopyNetworksButton () {
 }
 
+/**
+ * "Manage Storage Connections. Enabled when exactly 1 iSCSI domain is selected
```
```suggestion
 * Manage Storage Connections. Enabled when exactly 1 iSCSI domain is selected
```

mkemel replied:
Rephrased the comment

bennyz commented on src/modals/storage-connections/StorageConnectionsTable.js:
```diff
@@ -0,0 +1,356 @@
+}
+
+const canAttachConnection = (connection, storageDomain) => {
+  if (connection) {
+    return !connection.isAttachedToDomain && storageDomain.status === 'maintenance'
+  } else {
+    return false
+  }
+}
```
is it possible to do:
```suggestion
    return connection && !connection.isAttachedToDomain && storageDomain.status === 'maintenance'
```

barpavel replied:
Same for "canEditConnection", "canDetachConnection" & "canRemoveConnection".
Another option is to fail first and then no need for "else":
if (!connection) {
  return false
}

bennyz commented on src/modals/storage-connections/StorageConnectionsTable.js:
```diff
@@ -0,0 +1,356 @@
+  }
+
+  const createIScsiCells = (connection, rowIndex) => {
+    let i = 0
```
maybe
```suggestion
    let cellIndex = 0
```

barpavel commented on src/intl/messages.js:
```diff
@@ -1114,6 +1114,132 @@ const messageDescriptors = {
+  storageConnectionsAddConnectionButton: {
+    id: 'storage.domains.connections.connection.add.button',
+    defaultMessage: 'Add',
+    description: 'Add Connection buton',
```
button

barpavel commented on src/intl/messages.js:
```diff
@@ -1114,6 +1114,132 @@ const messageDescriptors = {
+  storageConnectionsAttachConnectionButton: {
+    id: 'storage.domains.connections.connection.attach.button',
+    defaultMessage: 'Attach',
+    description: 'Attach Connection buton',
```
button

barpavel commented on src/intl/messages.js:
```diff
@@ -1114,6 +1114,132 @@ const messageDescriptors = {
+  storageConnectionsDetachConnectionButton: {
+    id: 'storage.domains.connections.connection.detach.button',
+    defaultMessage: 'Detach',
+    description: 'Detach Connection buton',
```
button

mkemel replied:
Oh, those copy-paste-spreading typos :)

barpavel commented on src/intl/messages.js:
```diff
@@ -1114,6 +1114,132 @@ const messageDescriptors = {
+  storageConnectionsManageButton: {
+    id: 'storage.domains.connections.buttonLabel',
+    defaultMessage: 'Connections',
+    description: 'label for Storage Connections management dialog button',
```
Some 'description's start with lowercase, some with uppercase.

barpavel commented on src/intl/messages.js:
```diff
@@ -1114,6 +1114,132 @@ const messageDescriptors = {
+
+  storageConnectionsDomainNotInMaintenanceWarning: {
+    id: 'storage.domains.connections.domain.maintenance.warning',
+    defaultMessage: 'Storage Domain is not in Mainetnance mode',
```
Mainetnance --> Maintenance

barpavel commented on src/intl/messages.js:
```diff
@@ -1114,6 +1114,132 @@ const messageDescriptors = {
+  storageConnectionsDomainNotInMaintenanceWarning: {
+    id: 'storage.domains.connections.domain.maintenance.warning',
+    defaultMessage: 'Storage Domain is not in Mainetnance mode',
+    description: 'Storage Domain is not in Mainetnance mode Warning',
```
Mainetnance --> Maintenance

barpavel commented on src/intl/messages.js:
```diff
@@ -1114,6 +1114,132 @@ const messageDescriptors = {
+  storageConnectionsDomainNotInMaintenanceWarningDetail: {
+    id: 'storage.domains.connections.domain.maintenance.warning.detail',
+    defaultMessage: 'Connections cannot be attached or detached, cannot edit attached connections',
+    description: 'Storage Domain is not in Mainetnance mode Warning Detail',
```
Mainetnance --> Maintenance

barpavel commented on src/modals/storage-connections/StorageConnectionsDataProvider.js:
```diff
@@ -0,0 +1,124 @@
+  }
+
+  if (!domainConnections || domainConnections.error) {
+    throw new Error('StorageConnectionsDataProvider: failed to fetch storage connections' +
```
Maybe, not sure:
"storage connections" --> "domain connections" OR "storage domain connections"
?

mkemel replied:
Here, domain connections is in the meaning of connections attached to the domain
In the next line it says 'for storage domain ' + storageDomainId
I think it's enough

bennyz commented on src/integrations/buttons.js:
```diff
@@ -140,6 +141,29 @@ function addHostCopyNetworksButton () {
+    isEnabled: function (selectedDomains) {
+      return (
+        selectedDomains.length === 1 &&
+        selectedDomains[0] &&
+        selectedDomains[0].type &&
```
how can this happen?

mkemel replied:
Just in case the server returns incorrect object - so that it doesn't access null's fields and throws exception

ljelinkova commented on src/modals/storage-connections/StorageConnectionsDataProvider.js:
```diff
@@ -0,0 +1,124 @@
+import { msg } from '_/intl-messages'
+
+const createConnection = async (connection) => {
+  const response = await enginePost(
```
Why not just `return await enginePost(...)`? The same for other post/delete api calls.

mkemel replied:
Good point

ljelinkova commented on src/modals/storage-connections/StorageConnectionsDataProvider.js:
```diff
@@ -0,0 +1,124 @@
+
+  const domainConnectionsIds = new Set(domainConnections.map(connection => connection.id))
+  const isConnectionAttachedToDomain = (conn) => { return domainConnectionsIds.has(conn.id) }
+  const isConnectionNotAttachedToDomain = (conn) => { return !isConnectionAttachedToDomain(conn) }
```
I think it is not used.

mkemel replied:
It is, line 81

ljelinkova commented on src/modals/storage-connections/StorageConnectionsDataProvider.js:
```diff
@@ -0,0 +1,124 @@
+  }
+
+  const domainConnectionsIds = new Set(domainConnections.map(connection => connection.id))
+  const isConnectionAttachedToDomain = (conn) => { return domainConnectionsIds.has(conn.id) }
```
It seems a bit like overhead to create a separate function for such a simple condition.

mkemel replied:
Yes, but the lines that these conditions are used are already seemed to me quite cumbersome, so I thought it would improve the readability of the code

ljelinkova reviewed (commented):
Please attach some screenshots of the dialog.

mkemel commented:
> Please attach some screenshots of the dialog.

Done :+1:

rszwajko commented on src/modals/storage-connections/StorageConnectionsDataProvider.js:
```diff
@@ -0,0 +1,119 @@
+  const isConnectionNotAttachedToDomain = (conn) => { return !isConnectionAttachedToDomain(conn) }
+  const allConnectionsByType = allConnections
+    .filter(connection => connection.type === storageDomain.storage.type)
+    .map((conn) => {
```
by convention `map` does not mutate the array - consider returning a new object i.e. ` { ...conn, isAttachedToDomain:  value}`

mkemel replied:
Ack

rszwajko commented on src/modals/storage-connections/StorageConnectionsDataProvider.js:
```diff
@@ -0,0 +1,119 @@
+      return conn
+    })
+  domainConnections = allConnectionsByType.filter(isConnectionAttachedToDomain)
+  allConnections = domainConnections.concat(allConnectionsByType.filter(isConnectionNotAttachedToDomain))
```
it seems to me that at this point `allConnection` should be simply `allConectionByType`

rszwajko commented on src/modals/storage-connections/StorageConnectionsModal.js:
```diff
@@ -0,0 +1,154 @@
+  storageDomain,
+  domainConnections,
+  allConnections,
+  onClose = () => { },
```
the callbacks seem required properties - if yes they don't require defaults (but should be marked in proptypes as required)

mkemel replied:
Ack

rszwajko commented on src/modals/storage-connections/StorageConnectionsModal.js:
```diff
@@ -0,0 +1,154 @@
+  const createTitle = () => {
+    let title = msg.storageConnectionsTitle()
+    if (storageDomain) {
+      title = title + ' - ' + storageDomain.name
```
avoid concatenating strings in the code - in this case create a parametrized  version `storageConnectionsTitle` label

mkemel replied:
Yes, exactly for loading. I thought that it would be good to have the domain name that we are editing somewhere.
@sjd78 Do you think it's redundant?

rszwajko commented on src/modals/storage-connections/StorageConnectionsModal.js:
```diff
@@ -0,0 +1,154 @@
+    try {
+      await doConnectionCreate(connection)
+      setShowAll(true)
+      doRefreshConnections()
```
since `doRefreshConnections` is always called it seems a part of data provider logic

mkemel replied:
@rszwajko but since it's a callback passed by DataProvider component - I don't really have access to it in StorageConnectionsDataProvider, so the only way to call it inside doConnectionCreate is to pass it as an argument there. Thought I might as well call it here, also I think it is modal logic - I perform an action and then refresh the data. 
Or am I missing something here?

rszwajko replied:
>  it's a callback passed by DataProvider component -
>

Yes and you can define the callback there, while creating StorageConnectionsDataProvider i.e. somwhere before
```
 return React.cloneElement(child, {
        storageDomain,
        domainConnections,
        allConnections,
        doConnectionCreate: createConnection,
       .....
```

> I think it is modal logic
>

Ultimately everything what's on the modal is modal logic:) However by using data provider we extract that part to the data provider.

Btw `setShowAll(true)` does not belong here as it should be tied with React lifecycle. By the time the first call will succeed the component may be already unmounted which results in errors.

mkemel replied:
Maybe if it's okay then I would leave it the way it is. In that case, what would I need to sort out? Ignore fetchInProgress to not display the spinner? If yes - should I ignore it during the initial load as well? I tried different fetch errors, and it looked okay, but it is possible that I am missing something

sjd78 replied:
Spinner component looks ok.

The current code in `StorageConnectionsDataProvider` will activate the spinner every time a fetch is in progress.  You would need to tweak the data provider to change that behavior.

rszwajko commented on src/modals/storage-connections/StorageConnectionsModal.js:
```diff
@@ -0,0 +1,154 @@
+    return title
+  }
+
+  const onConnectionCreate = async (connection) => {
```
@sjd78 `async` function as a callback looks risky to me, what do you think?

rszwajko replied:
@sjd78 everything is OK as long as you don't call methods tied to React lifecycle i.e. things like `setShowAll(true)` are IMHO allowed only inside synchronous callback.

sjd78 replied:
@rszwajko - It is basically the same as a `useEffect()` function, and we're ok using state setters in. An async call in a useEffect() could finish after the component is unmounted, same as these.  I don't consider it much of a risk at this point.

rszwajko commented on src/modals/storage-connections/StorageConnectionsTable.js:
```diff
@@ -0,0 +1,340 @@
+  applyCellEdits,
+  validateCellEdits,
+  cellWidth,
+} from '@patternfly/react-table'
```
In the new code the composable table is preferred - check https://www.patternfly.org/v4/components/table
The legacy table is model driven and adds unnecessary complexity.

@sjd78   since the code is ready I'm not sure if it's worth changing, what do you think?

mkemel replied:
I have looked at Table Composable. It seems that Editable Rows are not supported there, and they are very much required here.

rszwajko replied:
The approach to editable rows is different, more React-compliant. Check i.e. my PR that uses them https://github.com/oVirt/ovirt-web-ui/pull/1549

sjd78 replied:
> In the new code the composable table is preferred - check https://www.patternfly.org/v4/components/table The legacy table is model driven and adds unnecessary complexity.
> 
> @sjd78 since the code is ready I'm not sure if it's worth changing, what do you think?

Good idea, but depends on if the composable table is available on the version of `@patternfly/react-core` we currently have in the project.

mkemel replied:
> Check i.e. my PR that uses them https://github.com/oVirt/ovirt-web-ui/pull/1549

@rszwajko nice, but I find it very explicit. Is there a special reason that TableComposable doesn't have this implemented in a more generalized way, while the legacy Table does (with quirks though)?

> depends on if the composable table is available on the version of @patternfly/react-core we currently have in the project.

@sjd78 I think it is, but is it that bad to use legacy Table here? I would really prefer not to rewrite this whole thing :(

rszwajko replied:
@mkemel 
> nice, but I find it very explicit. Is there a special reason that TableComposable doesn't have this implemented in a more > generalized way, while the legacy Table does (with quirks though)?
> 
 
1. less code i.e. check `src/components/CreateVmWizard/steps/Networking.js` in the mentioned PR which is basically refactoring from legacy PF3 into composable PF4
2. no restrictions on widgets used in read/edit mode - legacy has basically only 2 hacked cell types (text and select) that need to fit into (informal) interface
3. no restrictions on validation and usage pattern
4. React-style prop-driven component without intermediate row model and all index-based logic

sjd78 replied:
@rszwajko Using the `TableComposable` would be nice.  @mkemel It is more explicit, but in a good way.  The model and cell renderers and cell editors are much more complex than just doing normal widgets in a table.  I just went through going form the legacy Table to `TableComposable` style and the edit logic stayed the same but the render complexity went down.

So @mkemel, do you have time to spend converting it over?  We will have both kinds of tables in the code base, but it would be preferred to have the current style in new code.  Your code would not need much rework.

mkemel replied:
Having a working example now it might not be that horrible to rework that.

rszwajko replied:
Ack

rszwajko commented on src/modals/storage-connections/StorageConnectionsDataProvider.js:
```diff
@@ -0,0 +1,119 @@
+    throw new Error('StorageConnectionsDataProvider: failed to fetch data center')
+  }
+
+  storageDomain = await engineGet(`api/datacenters/${dataCenter.id}/storagedomains/${storageDomainId}`)
```
we retrieve the storage domain here for the second time - why this is required?

mkemel replied:
@rszwajko It's another endpoint: api/datacenters/${datacenter_id}/storagedomains/${storagedomain_id} vs. api/storagedomains/${storagedomain_id}
Only here the domain status is available

rszwajko replied:
IMHO both endpoints ( [/api/storagedomains/{id}](http://ovirt.github.io/ovirt-engine-api-model/master/#services/storage_domains/methods/list) and [/api/datacenters/{id}/storagedomains/{id}](http://ovirt.github.io/ovirt-engine-api-model/master/#services/attached_storage_domains/methods/list)) should return the same - this looks like a bug to me that should be fixed. If not possible please add comment in the code (with bug URL).

sjd78 replied:
@rszwajko - Those 2 endpoints do not return the same stuff, and can't.  Some info about a storage domain can only be accessed through the lense of a data center.  ovirt-web-ui has to do the same kind of thing when pulling SD and DC details at login.

rszwajko replied:
@sjd78 I haven't dig into this but according to REST API spec they return the same object - http://ovirt.github.io/ovirt-engine-api-model/master/#types/storage_domain
If there are some technical restrictions then that should be reflected in the API - if not this is a bug to me.

mkemel commented on src/modals/storage-connections/StorageConnectionsDataProvider.js:
```diff
@@ -0,0 +1,119 @@
+const fetchData = async (storageDomainId) => {
+  const allConnectionsJson = await engineGet('api/storageconnections')
+  let storageDomain = await engineGet(`api/storagedomains/${storageDomainId}`)
+  const domainConnectionsJson = await engineGet(`api/storagedomains/${storageDomainId}/storageconnections`)
```
@rszwajko now that you say that - I recall that I tried `follow` and saw that it doesn't work

rszwajko replied:
@mkemel please add a comment in the code about this  (if possible with link to  the bug) to avoid this discussion in the future:)

sjd78 replied:
Note: if multiple calls are done independently, you can run them all at once in parallel like this:

```js
const [ result1, result2, result3 ] = await Promise.all([
  fetch('something'),
  fetch('thing2'),
  fetch('thing3'),
])
```

sjd78 replied:
Note 2: you can also use the parallel promise call like that to introduce a loading time for testing/dev purposes.

For example, forcing the fetches to take at least 15 seconds so you can see the loading component, but only if you use the FAKE_DATA build (for example: `FAKE_DATA=true yarn dev`):
```jsx
import config from '_/plugin-config'

async function fetchData() {
  const [data, thing2, thing3] = await Promise.all([
    fetch('something'),
    fetch('thing2'),
    fetch('thing3'),
    config.useFakeData ? new Promise(resolve => setTimeout(() => { resolve() }, 15000)) : true,
  ])

  // further process `data` as needed
  const filteredAndById = data
    .filter(d => d.flag === 'purple')
    .map(d => { d.id, d.foo, d.bar, calcFooBar: d.fooBar === 'green' })
    .reduce((plucked, d) => { plucked[d.id] = d; return plucked }, {})

  return [filteredAndById, thing2, thing3]
}
```

sjd78 commented on src/modals/storage-connections/StorageConnectionsDataProvider.js:
```diff
@@ -0,0 +1,119 @@
+    throw new Error('StorageConnectionsDataProvider: failed to fetch storage domain')
+  }
+
+  let allConnections = allConnectionsJson.storage_connection
```
`eslint` should flag these...  maybe run a `yarn lint`?

Otherwise it is more unusual to see variables overwritten

mkemel replied:
I have run lint of course, it was not flagged, not even warning.
But I get it now that using const is not preferred. My first JS code :) Will change that

sjd78 replied:
Using `const` is preferred, but then the JS code will be more of a functional coding style.  Grab data, filter it, produce new output.

sjd78 reviewed (changes requested):
At least the state mismatch between the modal and modal body needs to be resolved.

I would prefer the table to follow the newer model, but it isn't a show stopper at this point.

The `fetchData` cross filtering is a bit awkward.  Does the `selectedDoamin` in `addManageStorageConnectionsButton()` have reference to the relevant data center id already?  If yes, pass that in and skip some of the lookups.

mkemel commented on src/modals/storage-connections/StorageConnectionsModalBody.js:
```diff
@@ -0,0 +1,120 @@
+              onClick={handleAddNewConnectionButtonClick}
+              isDisabled={isNewConnectionOpened}
+            >
+              {msg.storageConnectionsAddConnectionButton(handleAddNewConnectionButtonClick)}
```
Oops, might be code auto-complete

mkemel commented on src/modals/storage-connections/StorageConnectionsModal.js:
```diff
@@ -0,0 +1,154 @@
+  const [error, setError] = useState(null)
+  const [isShowAll, setShowAll] = useState(false)
+
+  const warning = (!storageDomain || storageDomain.status === 'maintenance')
```
Ack

mkemel commented:
@sjd78 @rszwajko 
1. Regarding moving to Table Composable - I have started to rewrite it and I understood that it would take me more time than I thought, especially for proofing the whole thing and making sure all the small logics are still working. So maybe leave it with the regular Table
2. Got rid of all the lets in fetchData. Also, added dataCenterId to the EntityObject, sparing one fetch out of 4
3. @sjd78 setShowAll - I need it  in StorageConnectionsModal to trigger it after successful Create action - new connections are naturally created unattached, and if only attached connections are shown we will not see the newly created connection. But I added useEffect to StorageConnectionsModal to sync the two states.

sjd78 commented on src/modals/storage-connections/StorageConnectionsDataProvider.js:
```diff
@@ -0,0 +1,121 @@
+        return null
+      }
+
+      if (fetchInProgress || !data) {
+        return React.cloneElement(child, { isLoading: true })
+      }
```
This will activate the spinner any time the `fetchAndUpdateData` is in progress (via `doRefreshConnections()` below.  To change it to only activate on the first fetch, do:
```suggestion
      if (!data) {
        return React.cloneElement(child, { isLoading: true })
      }
```

sjd78 reviewed (changes requested):
The changes a look good.  A few more comment to address, but nothing I'd consider major.

To reply to your comment:

> 1.  Regarding moving to Table Composable - I have started to rewrite it and I understood that it would take me more time than I thought, especially for proofing the whole thing and making sure all the small logics are still working. So maybe leave it with the regular Table

Given the time constraints, it's ok to keep the table component the way it is.

>  2. Got rid of all the lets in fetchData. Also, added dataCenterId to the EntityObject, sparing one fetch out of 4

The fetch code is looking much better.  Good work!

>  3. @sjd78 setShowAll - I need it in StorageConnectionsModal to trigger it after successful Create action - new connections are naturally created unattached, and if only attached connections are shown we will not see the newly created connection. But I added useEffect to StorageConnectionsModalBody to sync the two states.

Something with how those interplay is bugging me as being overly complex.  I'm not sure how to unwind it so it only need to be in one place.  But...since it seems to work, it is not a blocker.

>  4. Callbacks, so we agree to leave them in StorageConnectionsModal? Then error and isShowAll are staying there?

Callbacks are ok where they are.  In ovirt-web-ui, they would be redux calls, do stuff async and then push the results to the redux store and flow back down to the components to render.  There are ways to do this without redux.  A complex context object, or managing more state in the DataProvider making that the controlling component are two possibilities.  Trade-offs as they are, the current way doesn't seem and better, or worse.

Same as with `isShowAll` and `error`.

In general, keeping state in one component and having the rest be controlled component makes things "simpler".  The dependencies across components on state and effects I think gets the behavior you need at the cost of complexity.  It seems to work, so it isn't a blocker for me.

rszwajko commented on src/modals/storage-connections/StorageConnectionsTable.js:
```diff
@@ -0,0 +1,340 @@
+  // This function validates every cell according to cell.props.rules
+  // returns map (rule name => array of cell names that failed validation)
+  const validateRow = (row) => {
+    const validationErrors = row.cells.reduce((errorsMap, cell, index) => {
+      const testValue = cell.props.editableValue === '' ? '' : cell.props.editableValue || cell.props.value
+      const cellErrors = cell.props.rules.reduce((failedRules, rule) => {
+        if (!rule.validator(testValue)) {
+          failedRules.push(rule.name)
+        }
+        return failedRules
+      }, [])
+      cellErrors.forEach(ruleName => {
+        if (!errorsMap[ruleName]) {
+          errorsMap[ruleName] = []
+        }
+        errorsMap[ruleName].push(cell.props.name)
+      })
+      return errorsMap
+    }, {})
+    return validationErrors
```
@sjd78 I agree that `reduce` is not always the best choice but in case it could work but fully functional flow needs to be used i.e.
```
row.cells
      .filter(({ editableValue }) => /* hasEditableValue */ true)
      .flatMap(({ name, editableValue, rules }) =>
        rules.filter(rule => !rule.validator(editableValue))
          .map(rule => ({ cell: name, rule: rule.name }))
      ).reduce((acc, { cell, rule }) => ({ ...acc, rule: [...acc[rule], cell] }), {})

```

sjd78 replied:
@rszwajko - My cognative load to read the original is as high as to read yours.  I like the for loops just to keep it simple.

rszwajko commented on src/modals/storage-connections/StorageConnectionsTable.js:
```diff
@@ -0,0 +1,340 @@
+      cells: createCells(connection, rowIndex),
+      connection: connection,
+      isHoverable: true,
+      disableActions: !areActionsEnabled,
```
this should depend from `isInMaintance` and from output from `canEditConnection()`

rszwajko commented on src/modals/storage-connections/StorageConnectionsTable.js:
```diff
@@ -0,0 +1,340 @@
+}
+
+const canEditConnection = (connection, storageDomain) => {
+  return connection && (!connection.isAttachedToDomain || storageDomain.status === 'maintenance')
```
is this condition correct? can you edit detached connection when not in maintenance?

mkemel commented:
@sjd78 @rszwajko @ahadas 
I have posted changes that Scott proposed, except for the Spinner one, where it broke the behavior - table is not refreshed correctly. To change it I guess I would need to do some more work, and I have already stretched the time I could do stuff here on Friday :)

Also, now I see Radek also has some proposals that might be a good thing to implement.
I'm not sure when do we need it to be merged. I can continue working on this on Sunday.

rszwajko commented:
@mkemel 
TLDR I don't have a rock solid bug scenario that would block this PR so I'm OK with adding improvements later on.

Note I couldn't fully click through the dialog because it doesn't support FAKE_DATA and I don't have iscsi hardware. However, I managed to run it with my NFS domain after few hacks :) The user experience was mixed i.e.  consider the screenshot below
1. I could edit multiple connections (what will happen on referesh?)
2. it's possible to edit and non-editable row and then  save it (failed on `undefined` in validators)
3. "show all' looks like a kind of filter that gets auto-enabled after "create"
![image](https://user-images.githubusercontent.com/64194103/160141732-9b6a1e5a-1553-4615-9257-e1f4bfc12df6.png)

sgratch commented:
@mkemel the ui-extensions build won't start before Tuesday, so you can continue working on that on Sunday.

mkemel commented:
@rszwajko 
1. On refresh non-saved commit gets discarded. I didn't find how to disable the edit button back then. I guess if I would use TableComposable it would not be a problem :)
2. Saving a non-editable row (just like saving without editing) just cancels the edit. I don't get any errors there
3. Yes, that was the intent. I wanted a dialog that shows mainly existing SD connections, as far as I understand, 'Edit' is the main scenario there. But for some actions we need to see also other connections. I'm not a UX person, but I thought that double pane would make the dialog over complicated, when most of the time the user would like just to edit the connection.

Please review my latest changes, there were some in DataProvider and other places. I think it looks better now

rszwajko commented on src/modals/storage-connections/StorageConnectionsModal.js:
```diff
@@ -0,0 +1,148 @@
+}) => {
+  const [isOpen, setOpen] = useState(true)
+  const [error, setError] = useState(null)
+  const [isShowAll, setShowAll] = useState(false)
```
In the current version it still feels in the wrong place. It should go into`StorageConnectionsModalBody`   next to
`const [isNewConnectionOpened, setNewConnectionOpened] = useState(false)`

rszwajko commented on src/modals/storage-connections/StorageConnectionsModal.js:
```diff
@@ -0,0 +1,148 @@
+  const [isShowAll, setShowAll] = useState(false)
+  const [shownConnections, setShownConnections] = useState(null)
+
+  useEffect(() => {
```
`useEffect` here is not needed and might introduce bugs. Instead in `StorageConnectionsModalBody` we can simply do 
```
 <StorageConnectionsTable
          connections={connections.filter(conn => showAll || conn.isAttachedToDomain)}
```
This should work because the React component tree is re-rendered with every prop (i.e. `connections`) or state (i.e. `showAll`) change.

rszwajko reviewed (changes requested):
`showAll` features needs some changes. Beside that OK for me.

sgratch commented:
/ost

mkemel commented:
@sjd78 @sgratch Pls don't forget to squash on merge :)

sjd78 reviewed (approved):
PR looks reasonable at this point.  Can be merged once OST passes.

sjd78 commented:
/ost

sjd78 commented:
> /ost

Original OST passed https://github.com/oVirt/ovirt-engine-ui-extensions/actions/runs/2053199027

The only changes since then are a merge conflict on `src/constants.js` and a minor fix on `src/utils/fetch.js`.

sjd78 commented:
/ost
